### PR TITLE
Feature/automatic mail migration

### PR DESCRIPTION
**Pull request by AxelPeter**

### CODE REVIEW

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.controller.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        migrateAccount () {
+            this.Emails.migrateAccountToDestinationAccount(this.email.domain, this.email.accountName, this.migrate.destinationService, this.migrate.destinationEmail, this.migrate.password)
+                .then(() => this.Alerter.success(this.$scope.tr("email_tab_modal_migrate_success"), this.$scope.alerts.dashboard))
+                .catch((err) => this.handleError(err))
+                .finally(() => this.$scope.resetAction());
+        }
+
+        // Check from email-domain-email-account-change-password.controller.js
+        checkPassword (input) {
+            input.$setValidity("passwordCheck", !!this.migrate.password && !/^\s/.test(this.migrate.password) && !/\s$/.test(this.migrate.password) && !this.migrate.password.match(/[ÂÃÄÀÁÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿ]/));
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.controller.js]: Does `ng-messages` & `ng-pattern` can help you in the same way?

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                                <!-- /One Service -->
+
+                                <!-- Multiple Services -->
+                                <div class="radio-inline"
+                                    data-ng-if="ctrl.availableServices.length > 1"
+                                    data-ng-repeat="serviceType in ctrl.availableServices">
+                                    <label>
+                                        <input id="migrateServiceType{{$index}}" type="radio" name="serviceType"
+                                            data-ng-model="ctrl.migrate.serviceType"
+                                            data-ng-value="serviceType"> {{::serviceType}}
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: Prefer `ng-bind`?

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+                            <!-- /Has available email address -->
+
+                            <!-- No available email address -->
+                            <div data-ng-if="!ctrl.migrate.destinationEmail">
+                                <div class="alert alert-info" role="alert">
+                                    <p data-i18n-static="email_tab_modal_migrate_alert_configure_me"></p>
+                                </div>
+
+                                <p data-ng-bind-html="tr('email_tab_modal_migrate_no_available_email', ctrl.migrate.destinationService)"></p>
+                                <p ng-if="ctrl.migrate.serviceType === 'EMAIL PRO'"><a href="/#/configuration/email_pro/{{::ctrl.migrate.destinationService}}?tab=ACCOUNT" target="_blank" class="btn btn-primary" data-i18n-static="email_tab_modal_migrate_order_new_emails_button"></a></p>
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: `ng-if` => `data-ng-if`.

Comment (antleblanc) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: Any reason to specify attribut `target="_blank"`?

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                                <p ng-if="ctrl.migrate.serviceType === 'HOSTED EXCHANGE'"><a href="/#/configuration/exchange_hosted/{{::ctrl.migrate.destinationService}}/{{::ctrl.migrate.destinationService}}?tab=ACCOUNT" target="_blank" class="btn btn-primary" data-i18n-static="email_tab_modal_migrate_order_new_emails_button"></a></p>
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: `ng-if` => `data-ng-if`.

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+                    </div>
+                    <!-- /Has Services -->
+
+                </form>
+            </div>
+
+            <!-- Loader -->
+            <div class="text-center" data-ng-if="ctrl.loading">
+                <div class="oui-loader oui-loader_s d-inline-block">
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: Existing a directive called `loader` (see: https://github.com/ovh-ux/ovh-manager-web/blob/master/client/app/components/loader/loader.directive.js)

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                        <div class="col-md-8">
+                            <p class="form-control-static" data-ng-bind="ctrl.migrate.destinationService"></p>
+                        </div>
+                    </div>
+                    <!-- /Migration Data -->
+
+                </div>
+
+                <!-- Guides Text -->
+                <p ng-if="ctrl.migrate.serviceType === 'EMAIL PRO'" data-i18n-static="email_tab_modal_migrate_configuration_guides_emailpro"></p>
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: `ng-if` => `data-ng-if`.

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                <p ng-if="ctrl.migrate.serviceType === 'HOSTED EXCHANGE'" data-i18n-static="email_tab_modal_migrate_configuration_guides_exchange"></p>
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: `ng-if` => `data-ng-if`.

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+                                data-ng-show="ctrl.migratePasswordForm.accountPasswordConfirm.$dirty && ctrl.isPasswordDefined() && !ctrl.isPasswordMatches()"></small>
+                        </div>
+                    </div>
+                    <!-- Confirm Password -->
+
+                </form>
+            </div>
+
+            <!-- Loader -->
+            <div class="text-center" data-ng-if="ctrl.loading">
```
Comment (antleblanc) [inline client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: Directive `loader` instead?

File: client/app/resources/i18n/email/Messages_fr_FR.xml [hunk 9]
```diff
@@ -315,6 +316,31 @@
...
    <translation id="email_tab_modal_update_redirection_error" qtlid="269838">Une erreur est survenue lors de la modification de la redirection :</translation>
 
+   <translation id="email_tab_modal_migrate_title">Migrer vers E-mail pro ou Hosted Exchange</translation>
+   <translation id="email_tab_modal_migrate_select_platform_text">Choisissez le service de destination de votre adresse e-mail dans la liste ci-dessous.</translation>
+   <translation id="email_tab_modal_migrate_account_email_label">Email source :</translation>
+   <translation id="email_tab_modal_migrate_service_type_label">Offre de destination :</translation>
+   <translation id="email_tab_modal_migrate_destination_service_label">Service de destination :</translation>
+   <translation id="email_tab_modal_migrate_destination_service_help">Attention, vous ne pouvez migrer qu’un seul compte à la fois.</translation>
+   <translation id="email_tab_modal_migrate_password_text">Veuillez entrer le mot de passe de l'e-mail source pour valider la migration :</translation>
```
Comment (antleblanc) [inline client/app/resources/i18n/email/Messages_fr_FR.xml]: `l'e-mail` => `l’e-mail`

File: client/app/resources/i18n/email/Messages_fr_FR.xml [hunk 10]
```diff
@@ -315,6 +316,31 @@
...
+   <translation id="email_tab_modal_migrate_order_new_emails_button">Commander des comptes</translation>
+   <translation id="email_tab_modal_migrate_available_email_selected">La migration de votre email se fera sur le compte <strong>{0}</strong>.</translation>
+   <translation id="email_tab_modal_migrate_remaining_available_emails">Après la migration, il vous restera <strong>{0}</strong> compte(s) libre(s).</translation>
+   <translation id="email_tab_modal_migrate_no_remaining_available_emails">Après la migration, il ne vous restera plus de compte libre.</translation>
+   <translation id="email_tab_modal_migrate_no_available_email">Vous n'avez plus de compte disponible sur votre service <strong>{0}</strong>. Choisissez un autre service dans la liste ou commandez un nouveau compte</translation>
+   <translation id="email_tab_modal_migrate_order_email_service">Commander une solution email</translation>
```
Comment (antleblanc) [inline client/app/resources/i18n/email/Messages_fr_FR.xml]: `email` => `e-mail`.

File: client/app/resources/i18n/email/Messages_fr_FR.xml [hunk 11]
```diff
@@ -315,6 +316,31 @@
...
+   <translation id="email_tab_modal_migrate_no_email_service">Vous n'avez pas de solution e-mail compatible pour la migration.</translation>
```
Comment (antleblanc) [inline client/app/resources/i18n/email/Messages_fr_FR.xml]: `n’avez`

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html (hunk 4 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: bad habit

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html (hunk 2 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: Why it's preferred ? I'll have to add a useless span for the same goal

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.controller.js (hunk 1 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.controller.js]: Already seen with Joffrey. It's a copypaste from the modal "Change password"

File: client/app/resources/i18n/email/Messages_fr_FR.xml (hunk 9 above)
Comment (AxelPeter) [reply client/app/resources/i18n/email/Messages_fr_FR.xml]: done

File: client/app/resources/i18n/email/Messages_fr_FR.xml (hunk 10 above)
Comment (AxelPeter) [reply client/app/resources/i18n/email/Messages_fr_FR.xml]: done

File: client/app/resources/i18n/email/Messages_fr_FR.xml (hunk 11 above)
Comment (AxelPeter) [reply client/app/resources/i18n/email/Messages_fr_FR.xml]: done

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html (hunk 7 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: done

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html (hunk 6 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: done

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html (hunk 5 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: Ok, didn't know. Done

File: client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html (hunk 8 above)
Comment (AxelPeter) [reply client/app/email-domain/email/account/migrate/email-domain-email-account-migrate.html]: done

File: client/app/email-domain/email-domain.service.js [hunk 12]
```diff
@@ -1045,5 +1045,69 @@ angular.module("services").service(
                 broadcast: "hosting.tabs.delegate.responders.refresh"
             });
         }
+
+        /**
+         * Obtain a list of available services
+         * @param {string} domain
+         * @param {string} accountName
```
[the remark this answers is not in the document]
Comment (AxelPeter) [reply client/app/email-domain/email-domain.service.js]: done

**Pull request merged**